### PR TITLE
feat: add callbacks to support custom transaction config

### PR DESCRIPTION
**Pull request by Abhikumar98**
# FEATURE

Add callbacks to support custom transaction config [LF-763]

## TL;DR

Developers can now customize the gas configuration of their on-chain transactions using a callback function. This allows for modification of gas limit, gas price, or priority fee before executing the transaction.

## DESCRIPTION

This feature enables developers to customize the gas configuration of their on-chain transactions using a callback function. When a transaction request is sent, the callback function returns the TransactionRequest object to the developer, who can modify the gas configuration such as max gas limit, gas price, or priority fee. Once the developer returns the modified TransactionRequest object, the transaction is executed with the updated configuration. 

This feature is useful for simulating transactions or setting custom gas limits and prices. It allows developers to fine-tune transaction requests based on their requirements.

## STEPS TO TEST

To test this feature, follow the steps below:

1. Add a callback function to the transaction request
2. Modify the TransactionRequest object to customize the gas configuration
3. Return the modified TransactionRequest object via the callback function
4. Execute the transaction with the updated configuration

## EXPECTED RESULTS

The transaction should be executed with the customized gas configuration provided by the developer.

## CHECKLIST

- [x] Performed a self-review of my code
- [x] Added comments to the code for better readability
- [x] Made sure that existing tests are still passing.

- [ ] Added tests that cover the new functionality

[LF-763]: https://lifi.atlassian.net/browse/LF-763?atlOrigin=eyJpIjoiNWRkNTljNzYxNjVmNDY3MDlhMDU5Y2ZhYzA5YTRkZjUiLCJwIjoiZ2l0aHViLWNvbS1KU1cifQ

### CODE REVIEW

File: src/types/internal.types.ts [hunk 1]
```diff
@@ -24,6 +25,9 @@ export interface ExecutionParams {
   step: Step
   statusManager: StatusManager
   settings: InternalExecutionSettings
+  customConfigCallback?: (
```
Comment (chybisov) [inline src/types/internal.types.ts:28]: Please, let's find a better self-descriptive name for it. This one doesn't tell us what it is for.

**Comment by chybisov** [comment]
Please link a task for it or describe why we make these changes.

File: src/allowance/checkAllowance.ts [hunk 2]
```diff
@@ -13,7 +14,8 @@ export const checkAllowance = async (
   statusManager: StatusManager,
   settings: InternalExecutionSettings,
   chain: Chain,
-  allowUserInteraction = false
+  allowUserInteraction = false,
+  configCallback?: (tx: TransactionRequest) => Promise<TransactionRequest>
```
Comment (Addminus) [inline src/allowance/checkAllowance.ts:18]: maybe approvalConfigCallback is a more descriptive name?

**Review by Addminus — CHANGES_REQUESTED** [review]

File: src/allowance/checkAllowance.ts (hunk 2 above)
Comment (chybisov) [reply src/allowance/checkAllowance.ts:18]: I guess this is not only for approval but also for contract calls. https://lifi.atlassian.net/browse/LF-763

Maybe `updateTransactionRequest`?

Comment (Abhikumar98) [reply src/allowance/checkAllowance.ts]: updated the variable name

File: src/types/internal.types.ts [hunk 3]
```diff
@@ -24,6 +25,9 @@ export interface ExecutionParams {
   step: Step
   statusManager: StatusManager
   settings: InternalExecutionSettings
+  updateTransactionRequest?: (
+    transactionRequest: TransactionRequest
+  ) => Promise<TransactionRequest>
```
Comment (chybisov) [inline src/types/internal.types.ts:30]: Does it make sense to have it as part of the `InternalExecutionSettings` like we have other hooks?

**Review by chybisov — CHANGES_REQUESTED** [review]

File: src/types/internal.types.ts (hunk 3 above)
Comment (Abhikumar98) [reply src/types/internal.types.ts]: Yes this already seems to be used at multiple places. I'm using this in my change.

File: src/allowance/utils.ts [hunk 4]
```diff
@@ -9,13 +10,19 @@ import { fetchDataUsingMulticall, MultiCallData } from '../utils/multicall'
...
   const erc20 = new Contract(tokenAddress, ERC20_ABI, signer) as ERC20Contract
 
   try {
-    const approved = await erc20.allowance(signerAddress, contractAddress)
+    const approved = await erc20.allowance(signerAddress, contractAddress, {
+      gasLimit: transactionRequest?.gasLimit,
+      gasPrice: transactionRequest?.gasPrice,
+      maxFeePerGas: transactionRequest?.maxFeePerGas,
+      maxPriorityFeePerGas: transactionRequest?.maxPriorityFeePerGas,
+    })
```
Comment (chybisov) [inline src/allowance/utils.ts:25]: Can we pass the whole object here like 

`const approved = await erc20.allowance(signerAddress, contractAddress, transactionRequest)`

**Review by chybisov — CHANGES_REQUESTED** [review]

**Comment by chybisov** [comment]
Thanks for adding PR template 🙂

File: src/allowance/utils.ts (hunk 4 above)
Comment (Abhikumar98) [reply src/allowance/utils.ts:25]: I only want to pick these gas related changes that the developer modifies.
Hence avoiding the usage of whole object.

There are variables like: `data` and `customData` which I assumed shouldn't be modified. Although the contract would already have these fields filled with the ongoing interaction.

**Review by chybisov — DISMISSED** [review]

**Comment by chybisov** [comment]
@Abhikumar98 please add tests 🙂

**Review by chybisov — APPROVED** [review]
